source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -1189,6 +1134,123 @@ bool ContextImpl::verifyCertChain(X509& leaf_cert, STACK_OF(X509) & intermediate
+    logSslErrorChain();
+    throw EnvoyException(
+        absl::StrCat("Failed to load certificate chain from ", cert_chain_file_path_));
+  }
+  // Read rest of the certificate chain.
+  while (true) {
+    bssl::UniquePtr<X509> cert(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
+    if (cert == nullptr) {
+      break;
+    }
+    if (!SSL_CTX_add_extra_chain_cert(ssl_ctx_.get(), cert.get())) {
+      throw EnvoyException(
+          absl::StrCat("Failed to load certificate chain from ", cert_chain_file_path_));
+    }
+    // SSL_CTX_add_extra_chain_cert() takes ownership.
+    cert.release();
+  }
+  // Check for EOF.
+  const uint32_t err = ERR_peek_last_error();
+  if (ERR_GET_LIB(err) == ERR_LIB_PEM && ERR_GET_REASON(err) == PEM_R_NO_START_LINE) {
+    ERR_clear_error();
+  } else {
+    throw EnvoyException(
+        absl::StrCat("Failed to load certificate chain from ", cert_chain_file_path_));
+  }
+}
+
+void TlsContext::loadPrivateKey(const std::string& data, const std::string& data_path,
+                                const std::string& password) {
+  bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(const_cast<char*>(data.data()), data.size()));
+  RELEASE_ASSERT(bio != nullptr, "");
+  bssl::UniquePtr<EVP_PKEY> pkey(
+      PEM_read_bio_PrivateKey(bio.get(), nullptr, nullptr,
+                              !password.empty() ? const_cast<char*>(password.c_str()) : nullptr));
+
+  if (pkey == nullptr || !SSL_CTX_use_PrivateKey(ssl_ctx_.get(), pkey.get())) {
+    throw EnvoyException(fmt::format("Failed to load private key from {}, Cause: {}", data_path,
+                                     Utility::getLastCryptoError().value_or("unknown")));
+  }
+
+  checkPrivateKey(pkey, data_path);
+}
+
+void TlsContext::loadPkcs12(const std::string& data, const std::string& data_path,
+                            const std::string& password) {
+  cert_chain_file_path_ = data_path;
+  bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(const_cast<char*>(data.data()), data.size()));
+  RELEASE_ASSERT(bio != nullptr, "");
+  bssl::UniquePtr<PKCS12> pkcs12(d2i_PKCS12_bio(bio.get(), nullptr));
+
+  EVP_PKEY* temp_private_key = nullptr;
+  X509* temp_cert = nullptr;
+  STACK_OF(X509)* ca_certificates = nullptr;
+  if (pkcs12 == nullptr ||
+      !PKCS12_parse(pkcs12.get(), !password.empty() ? const_cast<char*>(password.c_str()) : nullptr,
+                    &temp_private_key, &temp_cert, &ca_certificates)) {
+    logSslErrorChain();
+    throw EnvoyException(absl::StrCat("Failed to load pkcs12 from ", data_path));
+  }
+  if (ca_certificates) {
+    X509* ca_cert = nullptr;
+    while ((ca_cert = sk_X509_pop(ca_certificates)) != nullptr) {
+      //
+      // This transfers ownership to ssl_ctx therefore ca_cert does not need to be freed.
+      //
+      SSL_CTX_add_extra_chain_cert(ssl_ctx_.get(), ca_cert);
+    }
+    sk_X509_free(ca_certificates);
+  }
+  cert_chain_.reset(temp_cert);
```

Comment:
Should this and the next line happen immediately after `PKCS12_parse` to make it clear what the ownership of the memory is? I don't think ther would be any functional difference, just readabilty